lusidtools/cocoon/cocoon.py

```diff
@@ -901,11 +904,342 @@ async def _construct_batches(
+        The required mapping
+    file_type : str
+        The type of file e.g. transactions, instruments, holdings, quotes, portfolios
+    sync_batches : list
+        A list of the batches used to upload the data into LUSID.
+
+    Returns
+    -------
+    responses: list
+        A list to be appended to the ultimate response for load_from_data_frame.
+
+        Where data_type == "transaction", the returned list will contain objects with the structure:
+            {transaction_id: instrument_identifiers}
+
+        Where data_type == "holding", the returned list will be a unique list of instrument_identifiers.
+    """
+    if file_type == "transaction":
+        return _unmatched_ids_transactions(
+            api_factory=api_factory,
+            scope=scope,
+            data_frame=data_frame,
+            mapping_required=mapping_required,
+            sync_batches=sync_batches,
+        )
+    elif file_type == "holding":
+        return _unmatched_ids_holdings(
+            api_factory=api_factory, scope=scope, sync_batches=sync_batches,
+        )
+
+
+def _unmatched_ids_transactions(
+    api_factory: lusid.utilities.ApiClientFactory,
+    scope: str,
+    data_frame: pd.DataFrame,
+    mapping_required: dict,
+    sync_batches: list = None,
+):
+    """
+    This method identifies which instruments were not resolved with a transaction upload using load_from_data_frame.
+
+    Parameters
+    ----------
+    api_factory : lusid.utilities.ApiClientFactory api_factory
+        The api factory to use
+    scope : str
+        The scope of the resource to load the data into
+    data_frame : pd.DataFrame
+        The DataFrame containing the data
+    sync_batches : list
+        A list of the batches used to upload the data into LUSID.
+
+    Returns
+    -------
+    responses: list
+        A list to be appended to the ultimate response for load_from_data_frame.
+
+        The returned list will contain objects with the structure:
+            {transaction_id: instrument_identifiers}
+
+    """
+    # Extract a list of portfolio codes from the sync_batches
+    portfolio_codes = _extract_unique_portfolio_codes(sync_batches)
+
+    # Create empty list to hold transaction ids and instruments
+    unmatched_transactions = []
+
+    # For each portfolio, request the unmatched transactions from LUSID and append to the instantiated list
+    for portfolio_code in portfolio_codes:
+        unmatched_transactions.extend(
+            return_unmatched_transactions(
+                api_factory=api_factory, scope=scope, code=portfolio_code,
+            )
+        )
+
+    # With the upload dataframe, filter out any transactions from the list that were not part of this upload and return
+    return filter_unmatched_transactions(
+        data_frame=data_frame,
+        mapping_required=mapping_required,
+        unmatched_transactions=unmatched_transactions,
+    )
+
+
+def _extract_unique_portfolio_codes(sync_batches: list):
```

Comment:
We could add unitttests for these functions

Reply:
added